module/metrics/herocache.go

```diff
@@ -0,0 +1,163 @@
+package metrics
+
+import (
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/onflow/flow-go/module"
+)
+
+const subsystemHeroCache = "hero_cache"
+
+type HeroCacheCollector struct {
+	normalizedBucketSlotAvailableHistogram prometheus.Histogram
+
+	successfulReadCount   prometheus.Counter
+	unsuccessfulReadCount prometheus.Counter
+
+	successfulWriteCount   prometheus.Counter
+	unsuccessfulWriteCount prometheus.Counter
+
+	fullCapacityEntityEjectionCount prometheus.Counter
+	emergencyKeyEjectionCount       prometheus.Counter
+}
+
+type HeroCacheMetricsRegistrationFunc func(uint64) module.HeroCacheMetrics
+
+func NetworkReceiveCacheMetricsFactory(registrar prometheus.Registerer) *HeroCacheCollector {
+	return NewHeroCacheCollector(namespaceNetwork, ResourceNetworkingReceiveCache, registrar)
+}
+
+func NetworkDnsTxtCacheMetricsFactory(registrar prometheus.Registerer) *HeroCacheCollector {
+	return NewHeroCacheCollector(namespaceNetwork, ResourceNetworkingDnsTxtCache, registrar)
+}
+
+func NetworkDnsIpCacheMetricsFactory(registrar prometheus.Registerer) *HeroCacheCollector {
+	return NewHeroCacheCollector(namespaceNetwork, ResourceNetworkingDnsIpCache, registrar)
+}
+
+func CollectionNodeTransactionsCacheMetrics(registrar prometheus.Registerer, epoch uint64) *HeroCacheCollector {
+	return NewHeroCacheCollector(namespaceCollection, fmt.Sprintf("%s_%d", ResourceTransaction, epoch), registrar)
+}
+
+func NewHeroCacheCollector(nameSpace string, cacheName string, registrar prometheus.Registerer) *HeroCacheCollector {
+
+	normalizedBucketSlotAvailableHistogram := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: nameSpace,
+		Subsystem: subsystemHeroCache,
+		Buckets:   []float64{0, 0.05, 0.1, 0.25, 0.5, 0.75, 1},
```

Comment:
Is it worth adding a comment about how these floating point numbers work?

Here's my assumption about how they work and how they would relate to heroCache:

heroCache generally uses 16 slots per bucket and therefore the available slots per bucket will be a number ranging from zero to 16, i.e. 17 different possible values. Whereas, the floats represent 7 possible values / prometheus 'buckets' not to be confused with heroCache 'buckets' :-) Have these 7 possible values been carefully chosen and if so, should this choosing process be commented on too?

So for the zero to 17 available slots:
* zero will map to 0,
* 1 will map to 0.625, <-- will this hit or miss the 0.05 bucket?
* 2 will map to 0.125, <-- will this hit or miss the 0.1 bucket?
* 3 will map to 0.1875,
* 4 will map to 0.25, <-- will this hit or miss the 0.25 bucket?
...
* 17 will map to 1.

Is it intended that the available slots 0, 1 , and 2 get mapped to prometheus buckets 0, 0.05, and 0.1?

Another question: Since the difference between 17 possible values and 7 buckets is only 10 then why not have 17 buckets and match the default number of heroCache slots?

Reply:
Added documentation here: https://github.com/onflow/flow-go/pull/2093/commits/2af1b4de49cf63fdc968e8da2690ca8925d92205

> Another question: Since the difference between 17 possible values and 7 buckets is only 10 then why not have 17 buckets and match the default number of heroCache slots?

Each bucket adds a time series per collector. Currently, we have 6 buckets, and adding 11 more means 11 more time-series per mempool that utilizes HeroCache. Currently 4 mempools use `HeroCache` and hence it adds 44 extra timeseries. This can simply go beyond 100 more timeseries when we switch more mempools to `HeroCache`. Not a scalable solution.